src/transforms/kubernetes/watch_client.rs

```diff
@@ -0,0 +1,469 @@
+    use crate::{
+        dns::Resolver,
+        sources::kubernetes::test::{echo, Kube},
+        test_util::runtime,
+    };
+    use dirs;
+    use futures::{future::Future, stream::Stream};
+    use http::Uri;
+    use kube::config::Config;
+    use openssl::{
+        ssl::{SslConnector, SslFiletype, SslMethod, SslVerifyMode},
+        x509::X509,
+    };
+    use serde_yaml;
+    use snafu::{ResultExt, Snafu};
+    use std::{fs::File, path::PathBuf, str::FromStr, sync::mpsc::channel, time::Duration};
+    use uuid::Uuid;
+
+    /// Enviorment variable that can containa path to kubernetes config file.
+    const CONFIG_PATH: &str = "KUBECONFIG";
+
+    /// Loads configuration from local kubeconfig file, the same
+    /// one that kubectl uses.
+    /// None if such file doesn't exist.
+    fn load_kube_config() -> Option<Result<Config, KubeConfigLoadError>> {
+        let path = std::env::var(CONFIG_PATH)
+            .ok()
+            .map(PathBuf::from)
+            .or_else(|| dirs::home_dir().map(|home| home.join(".kube").join("config")))?;
+
+        let file = match File::open(path) {
+            Ok(file) => file,
+            Err(error) if error.kind() == std::io::ErrorKind::NotFound => return None,
+            Err(error) => {
+                return Some(Err(KubeConfigLoadError::FileError { source: error }));
+            }
+        };
+
+        Some(serde_yaml::from_reader(file).context(ParsingError))
+    }
+
+    #[derive(Debug, Snafu)]
+    pub enum KubeConfigLoadError {
+        #[snafu(display("Error opening Kubernetes config file: {}.", source))]
+        FileError { source: std::io::Error },
+        #[snafu(display("Error parsing Kubernetes config file: {}.", source))]
+        ParsingError { source: serde_yaml::Error },
+    }
+
+    impl ClientConfig {
+        // NOTE: Currently used only for tests, but can be later used in
+        //       other places, but then the unsupported feature should be
+        //       implemented.
+        //
+        /// Loads configuration from local kubeconfig file, the same
+        /// one that kubectl uses.
+        fn load_kube_config(resolver: Resolver) -> Option<Self> {
+            let config = load_kube_config()?.unwrap();
+            // Get current context
+            let context = &config
+                .contexts
+                .iter()
+                .find(|context| context.name == config.current_context)?
+                .context;
+            // Get current user
+            let user = &config
+                .auth_infos
+                .iter()
+                .find(|user| user.name == context.user)?
+                .auth_info;
+            // Get current cluster
+            let cluster = &config
+                .clusters
+                .iter()
+                .find(|cluster| cluster.name == context.cluster)?
+                .cluster;
+            // The not yet supported features
+            assert!(user.username.is_none(), "Not yet supported");
```

Comment:
could you explain why we need these asserts?

Reply:
To catch if `kube` configuration is using them during tests. If they are, we would need to implement them, and that would require expanding `ClientConfig` with a feature that is needed only during testing.